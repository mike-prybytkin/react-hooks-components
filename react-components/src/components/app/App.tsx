import React from 'react';
import { Route, Routes, Navigate } from 'react-router-dom';
import AboutUs from 'pages/about-us/about-us';
import Page404 from 'pages/404/404';
import Main from 'pages/main/main';
import Header from 'components/header/header';
import { IProductCard } from 'share/types';
import { AppState } from './types';
import mockText from 'mocks/text';
import Form from 'pages/form/form';

class App extends React.Component<unknown, AppState> {
  constructor(props: unknown) {
    super(props);
    this.state = {
      data: [],
      isLoading: false,
      currentData: [],
      fetchError: '',
    };
    this.onSearch = this.onSearch.bind(this);
  }

  async fetchCards() {
    const URL = 'https://fakestoreapi.com/products?limit=20';
    this.setState({ isLoading: true });
    try {
      const response = await fetch(URL);
      const cards: IProductCard[] = await response.json();
      this.setState({
        data: cards,
        isLoading: false,
      });
    } catch (error) {
      this.setState({
        fetchError: error,
      });
    }
  }

  componentDidMount() {
    this.fetchCards();
  }

  onSearch(cards: IProductCard[]) {
    this.setState({
      currentData: cards,
    });
  }

  render() {
    const { isLoading } = this.state;
    return (
      <div className="App" data-testid="app">
        <Header cards={this.state.data} onSearch={this.onSearch} />
        <main className="main">
          <Routes>
            <Route
              path="/"
              element={<Main cards={this.state.currentData} heading={mockText.headingMain} />}
            />
            <Route
              path="/about"
              element={
                <AboutUs heading={mockText.headingAboutUs} message={mockText.pageNotReady} />
              }
            />
            <Route
              path="/form"
              element={
                <Form
                  heading={mockText.headingForm}
                  noCreatedUserMessage={mockText.notCreatedUserInForm}
                />
              }
            />
            <Route
              path="/notFound"
              element={
                <Page404
                  heading={mockText.heading404}
                  backToHomelinkText={mockText.linkBackToHome}
                />
              }
            />
            <Route path="/*" element={<Navigate to="/notFound" />} />
          </Routes>
          {isLoading && <p className="main__message_loading">{mockText.loading}</p>}
          {/* Следующий компонент в третьей таске я выношу в отрисовку карточек. Трогать не буду */}
          {this.state.currentData.length === 0 && !this.state.isLoading && (
            <p className="main__message_not-found">{mockText.itemNotFound}</p>
          )}
        </main>
      </div>
    );
  }
}

export default App;
