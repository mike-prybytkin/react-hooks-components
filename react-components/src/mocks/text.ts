const mockText = {
  homeLink: 'home',
  aboutLink: 'about',
  cardTitle: 'Fjallraven',
  searchPlaceholder: 'Search...',
  linkBackToHome: 'Back to home page',
  heading404: 'Error 404!',
  headingAboutUs: 'About Us!',
  headingMain: 'Welcome to us!',
  pageNotReady: 'Sorry, page not ready yet...',
  loading: 'Loading ...',
  itemNotFound: 'Sorry, item not found...',
  headingForm: 'Create User!',
  notCreatedUserInForm: 'Sorry, no users have been created yet...',
  labelUserName: 'Name:',
  placeholderUserName: 'Your Name',
  labelUserBirthday: 'Birthday:',
  labelUserSalary: 'Your monthly income:',
  switcherGender: 'Gender:',
  switcherGenderMale: 'Male',
  switcherGenderFemale: 'Female',
  labelUserMailing: 'Receive mailing',
  labelFileUpload: 'Add your photo',
  createUserButton: 'Create user',
  productCardButton: 'Learn more',
  prevButton: 'prev',
  nextButton: 'next',
  userSex: 'sex:',
  userBDay: 'b-day:',
  userSalary: 'salary:',
  subscribed: 'Subscribed',
  unsubscribed: 'Unsubscribed',
  toMailings: 'to mailings!',
  fromMailings: 'from mailings...',
};

export default mockText;
