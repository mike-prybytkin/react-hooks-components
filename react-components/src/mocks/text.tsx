const mockText = {
  homeLink: 'home',
  aboutLink: 'about',
  cardTitle: 'Fjallraven',
  searchPlaceholder: 'Search...',
  linkBackToHome: 'Back to home page',
  heading404: 'Error 404!',
  headingAboutUs: 'About Us!',
  headingMain: 'Welcome to our store!',
  pageNotReady: 'Sorry, page not ready yet...',
  loading: 'Loading ...',
  itemNotFound: 'Sorry, item not found...',
  headingForm: 'Create User!',
  notCreatedUserInForm: 'Sorry, no users have been created yet...',
  labelUserName: 'Name:',
  placeholderUserName: 'Your Name',
  labelUserBirthday: 'Birthday:',
  labelUserSalary: 'Your monthly income:',
  switcherGender: 'Gender:',
  switcherGenderMale: 'Male',
  switcherGenderFemale: 'Female',
  labelUserMailing: 'Receive mailing',
  labelFileUpload: 'Add your photo',
  createUserButton: 'Create user',
};

export default mockText;
