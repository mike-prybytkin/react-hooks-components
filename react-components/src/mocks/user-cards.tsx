import { IUserCard } from 'share/types';
import { userPhoto } from './user-photo';

export const mockUserCards: IUserCard[] = [
  {
    name: 'Max',
    avatarPath: userPhoto.max,
    gender: 'Male',
    birthday: '12-08-1995',
    salary: 'less than 100$',
    mailing: true,
  },
  {
    name: 'Kate',
    avatarPath: userPhoto.kate,
    gender: 'Female',
    birthday: '01-12-2003',
    salary: '200$ - 299$',
    mailing: false,
  },
];
