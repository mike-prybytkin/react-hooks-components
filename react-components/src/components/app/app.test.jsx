import React from 'react';
import { render, screen } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import App from './app';
import { mockCards } from '../../mocks/cards';
import mockText from '../../mocks/text';

global.fetch = jest.fn(() =>
  Promise.resolve({
    json: () => Promise.resolve({ mockCards }),
  })
);

const setUp = () => render(<App />, { wrapper: BrowserRouter });

describe('App', () => {
  it('should correctly render app structure', async () => {
    const { findByTestId } = setUp();
    const appNode = await findByTestId('app');
    expect(appNode.children).toHaveLength(2);
  });

  it('should correctly render the heading of the homepage', async () => {
    setUp();
    expect(await screen.findByText(new RegExp(mockText.headingMain, 'i'))).toBeInTheDocument();
  });

  it('should create snapshot', () => {
    const { asFragment } = setUp();
    expect(asFragment()).toMatchSnapshot();
  });
});
