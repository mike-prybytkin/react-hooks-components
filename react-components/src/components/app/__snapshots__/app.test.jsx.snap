// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should create snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
    data-testid="app"
  >
    <header
      class="header"
    >
      <nav
        class="navbar"
      >
        <ul
          class="navbar__items"
        >
          <li
            class="navbar__item"
          >
            <a
              aria-current="page"
              class="navbar__link active"
              href="/"
            >
              Home
            </a>
          </li>
          <li
            class="navbar__item"
          >
            <a
              class="navbar__link"
              href="/about"
            >
              About
            </a>
          </li>
          <li
            class="navbar__item"
          >
            <a
              class="navbar__link"
              href="/form"
            >
              Form
            </a>
          </li>
        </ul>
      </nav>
      <div
        class="search-container"
      >
        <form
          class="search-form"
        >
          <span>
            üîç
          </span>
          <input
            class="search"
            placeholder="Search..."
            value=""
          />
        </form>
      </div>
    </header>
    <main
      class="main"
    >
      <h1>
        Welcome to our store!
      </h1>
      <div
        class="cards-wrapper"
      />
      <p
        class="main__message_loading"
      >
        Loading ...
      </p>
    </main>
  </div>
</DocumentFragment>
`;
